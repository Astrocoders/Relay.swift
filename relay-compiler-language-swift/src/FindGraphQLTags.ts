import * as path from 'path';
import { GraphQLTagFinder } from 'relay-compiler/lib/language/RelayLanguagePluginInterface';
import { execFileSync } from 'child_process';

const toolPath = path.join(__dirname, 'find-graphql-tags');

export const find: GraphQLTagFinder = (_text, filePath) => {
  const output = execFileSync(toolPath, [filePath]);
  return JSON.parse(output);
};
