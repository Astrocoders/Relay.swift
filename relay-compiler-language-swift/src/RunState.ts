export class RunState {
  private generatedTypes = new Set<string>();

  addGeneratedType(name: string) {
    this.generatedTypes.add(name);
  }

  hasGeneratedType(name: string): boolean {
    return this.generatedTypes.has(name);
  }

  reset() {
    this.generatedTypes.clear();
  }
}
