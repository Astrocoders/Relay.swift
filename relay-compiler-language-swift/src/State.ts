import { TypeGeneratorOptions } from 'relay-compiler/lib/language/RelayLanguagePluginInterface';
import { InputStructNode } from 'SwiftGeneratorDataTypes';

export interface State extends TypeGeneratorOptions {
  usedEnums: Record<string, string>;
  generatedInputObjects: Record<string, 'pending' | InputStructNode>;
}
