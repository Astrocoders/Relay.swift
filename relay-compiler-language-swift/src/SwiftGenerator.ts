import { TypeGenerator } from 'relay-compiler';

export const generate: TypeGenerator['generate'] = (
  _schema,
  _node,
  options
) => {
  return JSON.stringify({ customScalars: options.customScalars });
};

export const transforms = [];
