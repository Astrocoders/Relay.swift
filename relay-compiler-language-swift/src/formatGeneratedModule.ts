import { FormatModule } from 'relay-compiler';

export const formatGeneratedModule: FormatModule = ({
  moduleName,
  node,
  documentType,
  docText,
  concreteText,
  typeText,
  hash,
  sourceHash,
}) => {
  return `
/* moduleName = ${moduleName} */
/* node = ${JSON.stringify(node, null, 4)} */
/* documentType = ${documentType} */
/* docText = ${docText} */
/* concreteText = ${concreteText} */
/* typeText = ${typeText} */
/* hash = ${hash} */
/* sourceHash = ${sourceHash} */
`;
};
