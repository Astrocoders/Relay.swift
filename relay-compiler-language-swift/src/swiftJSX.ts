export interface SwiftElement {
  type: string | SwiftComponent;
  props: any;
}
export type SwiftChild = SwiftElement | string;
export type SwiftNode = null | undefined | SwiftChild | SwiftFragment;
export type SwiftFragment = SwiftNode[];
export type SwiftComponent<Props = any> = (props: Props) => SwiftElement | null;

export function swiftJSX(
  tagName: string | SwiftComponent,
  attrs: any,
  ...children: SwiftNode[]
): SwiftElement | null {
  return { type: tagName, props: { ...attrs, children } };
}

export const Fragment = 'swiftFragment';

export function renderSwift(element: SwiftElement): string {
  const renderer = new Renderer();
  _renderSwift(renderer, element);
  return renderer.text;
}

interface RenderContext {
  inline?: boolean;
  inProtocol?: boolean;
  inSwitch?: boolean;
}

function _renderSwift(
  renderer: Renderer,
  element: SwiftNode,
  context: RenderContext = {}
) {
  if (!element) {
    return;
  }

  if (Array.isArray(element)) {
    for (const node of element) {
      _renderSwift(renderer, node, context);
    }
    return;
  }

  if (typeof element === 'string') {
    if (context.inline) {
      renderer.append(element);
    } else {
      renderer.appendLine(element);
    }
    return;
  }

  if (typeof element.type === 'string') {
    switch (element.type) {
      case 'swiftFragment':
        return _renderSwift(renderer, element.props.children);
      case 'init':
        return renderSwiftInit(renderer, element.props);
      case 'function':
        return renderSwiftFunction(renderer, element.props);
      case 'paramdecl':
        return renderSwiftFunctionParam(renderer, element.props);
      case 'struct':
        return renderSwiftStruct(renderer, element.props);
      case 'var':
        return renderSwiftVar(renderer, element.props, context);
      case 'extension':
        return renderSwiftExtension(renderer, element.props);
      case 'protocol':
        return renderSwiftProtocol(renderer, element.props);
      case 'enum':
        return renderSwiftEnum(renderer, element.props);
      case 'case':
        return renderSwiftCase(renderer, element.props, context);
      case 'import':
        return renderSwiftImport(renderer, element.props);
      case 'available':
        return renderSwiftAvailable(renderer, element.props);
      case 'call':
        return renderSwiftCall(renderer, element.props);
      case 'param':
        return renderSwiftParam(renderer, element.props, context);
      case 'compilecheck':
        return renderSwiftCompileTimeCheck(renderer, element.props);
      case 'typealias':
        return renderSwiftTypealias(renderer, element.props);
      case 'literal':
        return renderSwiftLiteral(renderer, element.props, context);
      case 'switch':
        return renderSwiftSwitch(renderer, element.props);
    }
  } else {
    const rendered = element.type(element.props);
    if (!rendered) {
      return;
    }
    return _renderSwift(renderer, rendered, context);
  }
}

class Renderer {
  private lines: string[];
  private level: number;
  private shouldContinue: boolean;

  constructor() {
    this.lines = [];
    this.level = 0;
    this.shouldContinue = false;
  }

  appendLine(line: string) {
    if (this.shouldContinue) {
      this.append(line);
    } else {
      this.lines.push('    '.repeat(this.level) + line);
    }
  }

  append(text: string) {
    if (this.shouldContinue) {
      this.shouldContinue = false;
    }

    if (this.lines.length) {
      this.lines[this.lines.length - 1] += text;
    } else {
      this.appendLine(text);
    }
  }

  continueLine() {
    this.shouldContinue = true;
  }

  indent() {
    this.level++;
  }

  outdent() {
    this.level--;
  }

  get text(): string {
    return this.lines.join('\n');
  }
}

type AccessLevel = 'public' | 'internal' | 'fileprivate' | 'private';

function renderSwiftStruct(
  renderer: Renderer,
  {
    name,
    inherit = [],
    access,
    children,
  }: {
    name: string;
    inherit?: string[];
    access?: AccessLevel;
    children: SwiftNode[];
  }
) {
  renderer.appendLine(
    `${access ? `${access} ` : ''}struct ${name}${
      inherit.length ? `: ${inherit.join(', ')}` : ''
    } {`
  );
  renderer.indent();

  _renderSwift(renderer, children);

  renderer.outdent();
  renderer.appendLine('}');
}

function renderSwiftVar(
  renderer: Renderer,
  {
    name,
    type,
    isStatic = false,
    access,
    children,
  }: {
    name: string;
    type: string;
    isStatic?: boolean;
    access?: AccessLevel;
    children: SwiftNode[];
  },
  context: RenderContext = {}
) {
  renderer.appendLine(
    `${access ? `${access} ` : ''}${
      isStatic ? 'static ' : ''
    }var ${name}: ${type}`
  );

  if (context.inProtocol) {
    renderer.append(' { get }');
    return;
  }

  if (children.length) {
    renderer.append(' {');
    renderer.indent();

    _renderSwift(renderer, children, context);

    renderer.outdent();
    renderer.appendLine('}');
  }
}

function renderSwiftFunction(
  renderer: Renderer,
  {
    name,
    parameters = [],
    returns,
    isStatic = false,
    access,
    children,
  }: {
    name: string;
    parameters: SwiftNode[];
    returns?: string;
    isStatic?: boolean;
    access?: AccessLevel;
    children: SwiftNode[];
  }
) {
  renderer.appendLine(
    `${access ? `${access} ` : ''}${isStatic ? 'static ' : ''}func ${name}(`
  );
  for (let i = 0; i < parameters.length; i++) {
    if (i !== 0) {
      renderer.append(', ');
    }
    _renderSwift(renderer, parameters[i], { inline: true });
  }
  renderer.append(`)`);

  if (returns) {
    renderer.append(` -> ${returns}`);
  }

  renderer.append(` {`);
  renderer.indent();

  _renderSwift(renderer, children);

  renderer.outdent();
  renderer.appendLine('}');
}

function renderSwiftInit(
  renderer: Renderer,
  {
    parameters = [],
    access,
    throws = false,
    children,
  }: {
    parameters: SwiftNode[];
    access?: AccessLevel;
    throws: boolean;
    children: SwiftNode[];
  }
) {
  renderer.appendLine(`${access ? `${access} ` : ''}init(`);
  for (let i = 0; i < parameters.length; i++) {
    if (i !== 0) {
      renderer.append(', ');
    }
    _renderSwift(renderer, parameters[i], { inline: true });
  }
  renderer.append(')');
  if (throws) {
    renderer.append(' throws');
  }
  renderer.append(' {');
  renderer.indent();

  _renderSwift(renderer, children);

  renderer.outdent();
  renderer.appendLine('}');
}

function renderSwiftFunctionParam(
  renderer: Renderer,
  {
    label,
    name,
    type,
    defaultValue,
  }: {
    label?: string;
    name: string;
    type: string;
    defaultValue?: SwiftNode;
  }
) {
  if (label) {
    renderer.append(`${label} `);
  }
  renderer.append(`${name}: ${type}`);
  if (defaultValue) {
    renderer.append(' = ');
    _renderSwift(renderer, defaultValue, { inline: true });
  }
}

function renderSwiftExtension(
  renderer: Renderer,
  {
    name,
    inherit = [],
    where = [],
    access,
    children,
  }: {
    name: string;
    inherit?: string[];
    where?: string[];
    access?: AccessLevel;
    children: SwiftNode[];
  }
) {
  renderer.appendLine(`${access ? `${access} ` : ''}extension ${name}`);
  if (inherit.length) {
    renderer.append(`: ${inherit.join(', ')}`);
  }
  if (where.length) {
    renderer.append(` where ${where.join(', ')}`);
  }
  renderer.append(' {');

  if (!children.length) {
    renderer.append('}');
    return;
  }

  renderer.indent();
  _renderSwift(renderer, children);
  renderer.outdent();
  renderer.appendLine('}');
}

function renderSwiftProtocol(
  renderer: Renderer,
  {
    name,
    access,
    children,
  }: {
    name: string;
    access?: AccessLevel;
    children: SwiftNode[];
  }
) {
  renderer.appendLine(`${access ? `${access} ` : ''}protocol ${name} {`);
  renderer.indent();

  _renderSwift(renderer, children, { inProtocol: true });

  renderer.outdent();
  renderer.appendLine('}');
}

function renderSwiftEnum(
  renderer: Renderer,
  {
    name,
    inherit = [],
    access,
    children,
  }: {
    name: string;
    inherit: string[];
    access?: AccessLevel;
    children: SwiftNode[];
  }
) {
  renderer.appendLine(`${access ? `${access} ` : ''}enum ${name}`);
  if (inherit.length) {
    renderer.append(`: ${inherit.join(', ')}`);
  }
  renderer.append(' {');
  renderer.indent();

  _renderSwift(renderer, children);

  renderer.outdent();
  renderer.appendLine('}');
}

function renderSwiftCase(
  renderer: Renderer,
  {
    name,
    parameters = [],
    isDefault = false,
    children,
  }: {
    name?: SwiftNode;
    parameters?: SwiftNode[];
    isDefault: boolean;
    children: SwiftNode[];
  },
  context: RenderContext = {}
) {
  if (isDefault) {
    renderer.appendLine('default');
  } else {
    renderer.appendLine('case ');
    _renderSwift(renderer, name, { inline: true });

    if (parameters.length) {
      renderer.append('(');

      for (let i = 0; i < parameters.length; i++) {
        if (i !== 0) {
          renderer.append(', ');
        }
        _renderSwift(renderer, parameters[i], { inline: true });
      }

      renderer.append(')');
    }
  }

  if (context.inSwitch) {
    renderer.append(':');
    renderer.indent();

    _renderSwift(renderer, children, context);

    renderer.outdent();
  } else {
    if (children.length) {
      renderer.append(' = ');
      _renderSwift(renderer, children, { inline: true });
    }
  }
}

function renderSwiftImport(
  renderer: Renderer,
  {
    module,
  }: {
    module: string;
  }
) {
  renderer.appendLine(`import ${module}`);
}

function renderSwiftAvailable(
  renderer: Renderer,
  {
    versions,
    children,
  }: {
    versions: string[];
    children: SwiftNode[];
  }
) {
  renderer.appendLine(`@available(${versions.join(', ')}, *)`);
  _renderSwift(renderer, children);
}

function renderSwiftCall(
  renderer: Renderer,
  {
    receiver,
    name,
    parameters = [],
    expanded = false,
  }: {
    receiver?: SwiftNode;
    name: string;
    parameters?: SwiftNode[];
    expanded?: boolean;
  }
) {
  renderer.appendLine(`${receiver != null ? `${receiver}.` : ''}${name}(`);
  if (expanded) {
    renderer.indent();
  }

  parameters = parameters.filter(p => !!p);
  for (let i = 0; i < parameters.length; i++) {
    if (i !== 0) {
      renderer.append(',');
      if (!expanded) {
        renderer.append(' ');
      }
    }
    _renderSwift(renderer, parameters[i], { inline: !expanded });
  }

  if (expanded) {
    renderer.outdent();
    renderer.appendLine(')');
  } else {
    renderer.append(')');
  }
}

function renderSwiftParam(
  renderer: Renderer,
  {
    label,
    children,
  }: {
    label?: string;
    children: SwiftNode[];
  },
  context: RenderContext = {}
) {
  const text = `${label ? `${label}: ` : ''}`;
  if (context.inline) {
    renderer.append(text);
  } else {
    renderer.appendLine(text);
  }

  renderer.continueLine();
  _renderSwift(renderer, children, { ...context, inline: true });
}

function renderSwiftCompileTimeCheck(
  renderer: Renderer,
  {
    condition,
    children,
  }: {
    condition: SwiftNode;
    children: SwiftNode[];
  }
) {
  renderer.appendLine('#if ');

  _renderSwift(renderer, condition, { inline: true });
  _renderSwift(renderer, children);

  renderer.appendLine('#endif');
}

function renderSwiftTypealias(
  renderer: Renderer,
  {
    name,
    children,
    access,
  }: {
    name: string;
    children: SwiftNode[];
    access?: AccessLevel;
  }
) {
  renderer.appendLine(`${access ? `${access} ` : ''}typealias ${name} = `);
  _renderSwift(renderer, children, { inline: true });
}

type SwiftLiteralProps =
  | { string: string }
  | { int: number }
  | { bool: boolean }
  | { array: SwiftNode[]; expanded?: boolean }
  | { dict: [SwiftNode, SwiftNode][]; expanded?: boolean };

function renderSwiftLiteral(
  renderer: Renderer,
  props: SwiftLiteralProps,
  context: RenderContext
) {
  const appendLine = context.inline
    ? text => {
        renderer.append(text);
      }
    : text => {
        renderer.appendLine(text);
      };

  if ('string' in props) {
    appendLine(JSON.stringify(props.string));
  } else if ('int' in props) {
    appendLine(String(props.int));
  } else if ('bool' in props) {
    appendLine(String(props.bool));
  } else if ('array' in props) {
    appendLine('[');
    renderer.indent();

    const innerContext = { ...context, inline: !props.expanded };

    for (let i = 0; i < props.array.length; i++) {
      if (i !== 0) {
        renderer.append(',');
        if (innerContext.inline) {
          renderer.append(' ');
        }
      }

      _renderSwift(renderer, props.array[i], innerContext);
    }

    renderer.outdent();
    if (props.expanded) {
      renderer.appendLine(']');
    } else {
      renderer.append(']');
    }
  } else if ('dict' in props) {
    appendLine('[');
    renderer.indent();

    const innerContext = { ...context, inline: !props.expanded };

    for (let i = 0; i < props.dict.length; i++) {
      if (i !== 0) {
        renderer.append(',');
        if (innerContext.inline) {
          renderer.append(' ');
        }
      }

      const [key, value] = props.dict[i];
      _renderSwift(renderer, key, innerContext);
      renderer.append(': ');
      renderer.continueLine();
      _renderSwift(renderer, value, innerContext);
    }

    renderer.outdent();
    if (props.expanded) {
      renderer.appendLine(']');
    } else {
      renderer.append(']');
    }
  } else {
    throw new Error('unrecognized props for literal');
  }
}

function renderSwiftSwitch(
  renderer: Renderer,
  {
    value,
    children,
  }: {
    value: SwiftNode;
    children: SwiftNode[];
  }
) {
  renderer.appendLine('switch ');
  _renderSwift(renderer, value, { inline: true });
  renderer.append(' {');

  _renderSwift(renderer, children, { inSwitch: true });

  renderer.appendLine('}');
}
