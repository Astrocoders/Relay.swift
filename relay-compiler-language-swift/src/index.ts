import { PluginInterface } from 'relay-compiler/lib/language/RelayLanguagePluginInterface';
import { find } from './FindGraphQLTags';
import { SwiftGenerator } from './SwiftGenerator';
import { formatGeneratedModule } from './formatGeneratedModule';
import { RunState } from './RunState';

import * as path from 'path';
import * as fs from 'fs';

export default function plugin(): PluginInterface {
  const runState = new RunState();
  return {
    inputExtensions: ['swift'],
    outputExtension: 'swift',
    getFileFilter(baseDir: string) {
      return (file: any) => {
        runState.reset();
        const filePath = path.join(baseDir, file.relPath);
        let text = '';
        try {
          text = fs.readFileSync(filePath, 'utf8');
        } catch {
          // eslint-disable no-console
          console.warn(
            `RelaySourceModuleParser: Unable to read the file "${filePath}". Looks like it was removed.`
          );
          return false;
        }
        return text.indexOf('graphql') >= 0;
      };
    },
    findGraphQLTags: find,
    formatModule: formatGeneratedModule(),
    typeGenerator: SwiftGenerator(runState),
  } as any;
}
