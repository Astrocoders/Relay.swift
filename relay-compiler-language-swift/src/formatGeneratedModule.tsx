/** @jsx swiftJSX */

import { FormatModule } from 'relay-compiler';
import { swiftJSX, renderSwift, DeclarationGroup } from './swiftJSX';
import { NodeStruct } from './components/NodeStruct';

export function formatGeneratedModule(): FormatModule {
  return ({ node, typeText }) => {
    return renderSwift(
      <DeclarationGroup>
        {`// Auto-generated by relay-compiler. Do not edit.`}
        <import module="Relay" />
        <NodeStruct node={node} typeText={typeText} />
      </DeclarationGroup>
    );
  };
}
