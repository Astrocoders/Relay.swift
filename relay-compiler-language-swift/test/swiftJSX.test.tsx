/** @jsx swiftJSX */

import { swiftJSX, renderSwift, Fragment } from '../src/swiftJSX';

test('struct with some vars', () => {
  const code = (
    <struct name="Data">
      <var name="foo" type="String" />
      <var name="bar" type="Int?" />
    </struct>
  );

  expect(renderSwift(code)).toMatchSnapshot();
});

test('struct with a nested struct', () => {
  const code = (
    <struct name="Data" inherit={['Decodable', 'Identifiable']}>
      <var name="tweets" type="[Tweet_tweets]" />
      <var name="id" type="String" />
      <var name="status" type="TweetStatus" />

      <struct name="Tweet_tweets" inherit={['Decodable']}>
        <var name="postedTweetID" type="String?" />
      </struct>
    </struct>
  );

  expect(renderSwift(code)).toMatchSnapshot();
});

test('struct with an initializer', () => {
  const code = (
    <struct name="DetailedTweetRow_tweet">
      <init
        parameters={[
          <paramdecl name="key" type="DetailedTweetRow_tweet_Key" />,
        ]}
      >
        fragmentPointer = key.fragment_DetailedTweetRow_tweet
      </init>
    </struct>
  );

  expect(renderSwift(code)).toMatchSnapshot();
});

test('empty extension', () => {
  const code = (
    <extension name="DetailedTweetRow_tweet" inherit={['Relay.Fragment']} />
  );

  expect(renderSwift(code)).toMatchSnapshot();
});

test('extension with nested types', () => {
  const code = (
    <extension name="DetailedTweetRow_tweet">
      <struct name="Data" inherit={['Decodable', 'Identifiable']}>
        <var name="tweets" type="[Tweet_tweets]" />
        <var name="id" type="String" />
        <var name="status" type="TweetStatus" />

        <struct name="Tweet_tweets" inherit={['Decodable']}>
          <var name="postedTweetID" type="String?" />
        </struct>
      </struct>
    </extension>
  );

  expect(renderSwift(code)).toMatchSnapshot();
});

test('fragment with an import', () => {
  const code = (
    <Fragment>
      <import module="Relay" />
      <struct name="Data">
        <var name="foo" type="String" />
        <var name="bar" type="Int?" />
      </struct>
    </Fragment>
  );

  expect(renderSwift(code)).toMatchSnapshot();
});

test('function tagged with availability', () => {
  const code = (
    <extension name="TweetRow_tweetGroup_Key">
      <available
        versions={['iOS 14.0', 'macOS 10.16', 'tvOS 14.0', 'watchOS 7.0']}
      >
        <function
          name="asFragment"
          returns="RelaySwiftUI.FragmentNext<TweetRow_tweetGroup>"
        >
          {'RelaySwiftUI.FragmentNext<TweetRow_tweetGroup>(self)'}
        </function>
      </available>
    </extension>
  );

  expect(renderSwift(code)).toMatchSnapshot();
});

test('public access', () => {
  const code = (
    <Fragment>
      <extension
        name="TweetRow_tweetGroup_Key"
        inherit={['Relay.Operation']}
        access="public"
      >
        <function
          name="asFragment"
          returns="RelaySwiftUI.FragmentNext<TweetRow_tweetGroup>"
          access="public"
        >
          {'RelaySwiftUI.FragmentNext<TweetRow_tweetGroup>(self)'}
        </function>
      </extension>
      <struct name="Data" access="public">
        <var name="foo" type="String" access="public" />
        <var name="bar" type="Int?" access="public" />
      </struct>
    </Fragment>
  );

  expect(renderSwift(code)).toMatchSnapshot();
});

test('property with a function call expression', () => {
  const code = (
    <extension name="TweetList_tweets" inherit={['Relay.PaginationFragment']}>
      <var name="metadata" type="Metadata" isStatic>
        <call
          name="RefetchMetadata"
          parameters={[
            <param label="path">["viewer"]</param>,
            <param label="operation">Operation.self</param>,
            <param label="connection">
              <call
                name="ConnectionMetadata"
                parameters={[
                  <param label="path">["allTweets"]</param>,
                  <param label="forward">
                    <call
                      name="ConnectionVariableConfig"
                      parameters={[
                        <param label="count">"count"</param>,
                        <param label="cursor">"cursor"</param>,
                      ]}
                    />
                  </param>,
                ]}
                expanded
              />
            </param>,
          ]}
          expanded
        />
      </var>
    </extension>
  );

  expect(renderSwift(code)).toMatchSnapshot();
});

test('nested components', () => {
  const ConnectionVariableConfig = ({ count, cursor }) => {
    return (
      <call
        name="ConnectionVariableConfig"
        parameters={[
          <param label="count">"{count}"</param>,
          <param label="cursor">"{cursor}"</param>,
        ]}
      />
    );
  };

  const ConnectionMetadata = ({
    path,
    isForward = false,
    isBackward = false,
  }) => {
    return (
      <call
        name="ConnectionMetadata"
        parameters={[
          <param label="path">[{path.map(s => `"${s}"`).join(', ')}]</param>,
          isForward ? (
            <param label="forward">
              <ConnectionVariableConfig count="count" cursor="cursor" />
            </param>
          ) : null,
          isBackward ? (
            <param label="backward">
              <ConnectionVariableConfig count="count" cursor="cursor" />
            </param>
          ) : null,
        ]}
        expanded
      />
    );
  };

  const RefetchMetadata = ({ path, connectionPath }) => {
    return (
      <call
        name="RefetchMetadata"
        parameters={[
          <param label="path">[{path.map(s => `"${s}"`).join(', ')}]</param>,
          <param label="operation">Operation.self</param>,
          <param label="connection">
            <ConnectionMetadata path={connectionPath} isForward />
          </param>,
        ]}
        expanded
      />
    );
  };

  const code = (
    <extension name="TweetList_tweets" inherit={['Relay.PaginationFragment']}>
      <var name="metadata" type="Metadata" isStatic>
        <RefetchMetadata path={['viewer']} connectionPath={['allTweets']} />
      </var>
    </extension>
  );

  expect(renderSwift(code)).toMatchSnapshot();
});
