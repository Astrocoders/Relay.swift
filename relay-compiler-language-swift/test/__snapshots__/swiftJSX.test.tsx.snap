// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`empty extension 1`] = `"extension DetailedTweetRow_tweet: Relay.Fragment {}"`;

exports[`extension with nested types 1`] = `
"extension DetailedTweetRow_tweet {
    struct Data: Decodable, Identifiable {
        var tweets: [Tweet_tweets]
        var id: String
        var status: TweetStatus
        struct Tweet_tweets: Decodable {
            var postedTweetID: String?
        }
    }
}"
`;

exports[`fragment with an import 1`] = `
"import Relay
struct Data {
    var foo: String
    var bar: Int?
}"
`;

exports[`function tagged with availability 1`] = `
"extension TweetRow_tweetGroup_Key {
    @available(iOS 14.0, macOS 10.16, tvOS 14.0, watchOS 7.0, *)
    func asFragment() -> RelaySwiftUI.FragmentNext<TweetRow_tweetGroup> {
        RelaySwiftUI.FragmentNext<TweetRow_tweetGroup>(self)
    }
}"
`;

exports[`nested components 1`] = `
"extension TweetList_tweets: Relay.PaginationFragment {
    static var metadata: Metadata {
        RefetchMetadata(
            path: [\\"viewer\\"],
            operation: Operation.self,
            connection: ConnectionMetadata(
                path: [\\"allTweets\\"],
                forward: ConnectionVariableConfig(count: \\"count\\", cursor: \\"cursor\\")
            )
        )
    }
}"
`;

exports[`property with a function call expression 1`] = `
"extension TweetList_tweets: Relay.PaginationFragment {
    static var metadata: Metadata {
        RefetchMetadata(
            path: [\\"viewer\\"],
            operation: Operation.self,
            connection: ConnectionMetadata(
                path: [\\"allTweets\\"],
                forward: ConnectionVariableConfig(count: \\"count\\", cursor: \\"cursor\\")
            )
        )
    }
}"
`;

exports[`public access 1`] = `
"public extension TweetRow_tweetGroup_Key: Relay.Operation {
    public func asFragment() -> RelaySwiftUI.FragmentNext<TweetRow_tweetGroup> {
        RelaySwiftUI.FragmentNext<TweetRow_tweetGroup>(self)
    }
}
public struct Data {
    public var foo: String
    public var bar: Int?
}"
`;

exports[`struct with a nested struct 1`] = `
"struct Data: Decodable, Identifiable {
    var tweets: [Tweet_tweets]
    var id: String
    var status: TweetStatus
    struct Tweet_tweets: Decodable {
        var postedTweetID: String?
    }
}"
`;

exports[`struct with an initializer 1`] = `
"struct DetailedTweetRow_tweet {
    init(key: DetailedTweetRow_tweet_Key) {
        fragmentPointer = key.fragment_DetailedTweetRow_tweet
    }
}"
`;

exports[`struct with some vars 1`] = `
"struct Data {
    var foo: String
    var bar: Int?
}"
`;
